File: plc4go/internal/spi/utils/ReadBufferXmlBased.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package utils

import (
	"encoding/hex"
	"encoding/xml"
	"fmt"
	"github.com/pkg/errors"
	"io"
	"math/big"
	"strings"
)

// NewXmlReadBuffer return as ReadBuffer which doesn't validate attributes and lists
func NewXmlReadBuffer(reader io.Reader) ReadBuffer {
	return &xmlReadBuffer{
		Decoder:        xml.NewDecoder(reader),
		pos:            1,
		doValidateList: false,
		doValidateAttr: false,
	}
}

// NewStrictXmlReadBuffer return as ReadBuffer which does validate attributes and lists depending on the setting
func NewStrictXmlReadBuffer(reader io.Reader, validateAttr bool, validateList bool) ReadBuffer {
	return &xmlReadBuffer{
		Decoder:        xml.NewDecoder(reader),
		pos:            1,
		doValidateAttr: validateAttr,
		doValidateList: validateList,
	}
}

///////////////////////////////////////
///////////////////////////////////////
//
// Internal section
//

type xmlReadBuffer struct {
	bufferCommons
	*xml.Decoder
	pos            uint
	doValidateAttr bool
	doValidateList bool
}

//
// Internal section
//
///////////////////////////////////////
///////////////////////////////////////

func (x *xmlReadBuffer) GetPos() uint16 {
	return uint16(x.pos / 8)
}

func (x *xmlReadBuffer) Reset(pos uint16) {
	x.pos = uint(pos * 8)
}

func (x *xmlReadBuffer) HasMore(bitLength uint8) bool {
	// TODO: work with x.InputOffset() and check if we are at EOF
	return true
}

func (x *xmlReadBuffer) PullContext(logicalName string, readerArgs ...WithReaderArgs) error {
	startElement, err := x.travelToNextStartElement()
	if err != nil {
		return err
	}
	if startElement.Name.Local != logicalName {
		return errors.Errorf("Unexpected Start element '%s'. Expected '%s'", startElement.Name.Local, logicalName)
	}
	if err := x.validateIfList(readerArgs, startElement); err != nil {
		return err
	}
	return nil
}

func (x *xmlReadBuffer) ReadBit(logicalName string, readerArgs ...WithReaderArgs) (bool, error) {
	var value bool
	err := x.decode(logicalName, rwBitKey, 1, readerArgs, &value)
	if err != nil {
		return false, err
	}
	x.move(1)
	return value, nil
}

func (x *xmlReadBuffer) ReadByte(logicalName string, readerArgs ...WithReaderArgs) (byte, error) {
	var value string
	err := x.decode(logicalName, rwByteKey, 8, readerArgs, &value)
	if err != nil {
		return 0, err
	}
	hexString := value
	if !strings.HasPrefix(hexString, "0x") {
		return 0, errors.Errorf("Hex string should start with 0x. Actual value %s", hexString)
	}
	hexString = strings.Replace(hexString, "0x", "", 1)
	decoded, err := hex.DecodeString(hexString)
	if err != nil {
		return 0, err
	}
	x.move(8)
	return decoded[0], nil
}

func (x *xmlReadBuffer) ReadByteArray(logicalName string, numberOfBytes int, readerArgs ...WithReaderArgs) ([]byte, error) {
	var value string
	err := x.decode(logicalName, rwByteKey, uint(numberOfBytes/8), readerArgs, &value)
	if err != nil {
		return nil, err
	}
	hexString := value
	if !strings.HasPrefix(hexString, "0x") {
		return nil, errors.Errorf("Hex string should start with 0x. Actual value %s", hexString)
	}
	hexString = strings.Replace(hexString, "0x", "", 1)
	decoded, err := hex.DecodeString(hexString)
	if err != nil {
		return nil, err
	}
	x.move(uint8(numberOfBytes / 8))
	return decoded, nil
}

func (x *xmlReadBuffer) ReadUint8(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (uint8, error) {
	var value uint8
	err := x.decode(logicalName, rwUintKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadUint16(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (uint16, error) {
	var value uint16
	err := x.decode(logicalName, rwUintKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadUint32(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (uint32, error) {
	var value uint32
	err := x.decode(logicalName, rwUintKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadUint64(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (uint64, error) {
	var value uint64
	err := x.decode(logicalName, rwUintKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadInt8(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (int8, error) {
	var value int8
	err := x.decode(logicalName, rwIntKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadInt16(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (int16, error) {
	var value int16
	err := x.decode(logicalName, rwIntKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadInt32(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (int32, error) {
	var value int32
	err := x.decode(logicalName, rwIntKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadInt64(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (int64, error) {
	var value int64
	err := x.decode(logicalName, rwIntKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadBigInt(logicalName string, bitLength uint64, readerArgs ...WithReaderArgs) (*big.Int, error) {
	var value big.Int
	err := x.decode(logicalName, rwIntKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return nil, err
	}
	x.move(uint8(bitLength))
	return &value, nil
}

func (x *xmlReadBuffer) ReadFloat32(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (float32, error) {
	var value float32
	err := x.decode(logicalName, rwFloatKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadFloat64(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (float64, error) {
	var value float64
	err := x.decode(logicalName, rwFloatKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return 0, err
	}
	x.move(bitLength)
	return value, nil
}

func (x *xmlReadBuffer) ReadBigFloat(logicalName string, bitLength uint8, readerArgs ...WithReaderArgs) (*big.Float, error) {
	var value big.Float
	err := x.decode(logicalName, rwFloatKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return nil, err
	}
	x.move(bitLength)
	return &value, nil
}

func (x *xmlReadBuffer) ReadString(logicalName string, bitLength uint32, readerArgs ...WithReaderArgs) (string, error) {
	var value string
	// TODO: bitlength too short
	err := x.decode(logicalName, rwStringKey, uint(bitLength), readerArgs, &value)
	if err != nil {
		return "", err
	}
	x.move(uint8(bitLength))
	return value, nil
}

func (x *xmlReadBuffer) CloseContext(logicalName string, _ ...WithReaderArgs) error {
	endElement, err := x.travelToNextEndElement()
	if err != nil {
		return err
	}
	if endElement.Name.Local != logicalName {
		return errors.Errorf("Unexpected End element '%s'. Expected '%s'", endElement.Name.Local, logicalName)
	}
	return nil
}

func (x *xmlReadBuffer) move(bits uint8) {
	x.pos += uint(bits)
}

func (x *xmlReadBuffer) travelToNextStartElement() (xml.StartElement, error) {
	for {
		token, err := x.Token()
		if err != nil {
			return xml.StartElement{}, err
		}
		switch token.(type) {
		case xml.StartElement:
			return token.(xml.StartElement), nil
		case xml.EndElement:
			return xml.StartElement{}, errors.Errorf("unexpected end element %s", token.(xml.EndElement).Name)
		}
	}
}

func (x *xmlReadBuffer) travelToNextEndElement() (xml.EndElement, error) {
	var endElement xml.EndElement
findTheEndToken:
	for {
		token, err := x.Token()
		if err != nil {
			return xml.EndElement{}, err
		}
		switch token.(type) {
		case xml.EndElement:
			endElement = token.(xml.EndElement)
			break findTheEndToken
		case xml.StartElement:
			return xml.EndElement{}, errors.Errorf("unexpected start element %s", token.(xml.StartElement).Name)
		}
	}
	return endElement, nil
}

func (x *xmlReadBuffer) decode(logicalName string, dataType string, bitLength uint, readerArgs []WithReaderArgs, targetValue interface{}) error {
	startElement, err := x.travelToNextStartElement()
	if err != nil {
		return err
	}
	err = x.validateStartElement(startElement, logicalName, dataType, bitLength, readerArgs...)
	if err != nil {
		return err
	}
	err = x.DecodeElement(targetValue, &startElement)
	if err != nil {
		return err
	}
	return nil
}

func (x *xmlReadBuffer) validateIfList(readerArgs []WithReaderArgs, startElement xml.StartElement) error {
	if !x.doValidateList {
		return nil
	}
	if x.isToBeRenderedAsList(upcastReaderArgs(readerArgs...)...) {
		for _, attr := range startElement.Attr {
			switch attr.Name.Local {
			case rwIsListKey:
				if attr.Value != "true" {
					return errors.Errorf("Startelement should be marked as %s=true", rwIsListKey)
				}
			}
		}
	}
	return nil
}

func (x *xmlReadBuffer) validateStartElement(startElement xml.StartElement, logicalName string, dataType string, bitLength uint, readerArgs ...WithReaderArgs) error {
	logicalName = x.sanitizeLogicalName(logicalName)
	if startElement.Name.Local != logicalName {
		return errors.Errorf("unexpected start element '%s'. Expected '%s'", startElement.Name.Local, logicalName)
	} else if err := x.validateAttr(startElement.Attr, dataType, bitLength, readerArgs...); err != nil {
		return errors.Wrap(err, "Error validating Attributes")
	}
	return nil
}

func (x *xmlReadBuffer) validateAttr(attr []xml.Attr, dataType string, bitLength uint, _ ...WithReaderArgs) error {
	if !x.doValidateAttr {
		return nil
	}
	dataTypeValidated := false
	bitLengthValidate := false
	for _, attribute := range attr {
		switch attribute.Name.Local {
		case rwDataTypeKey:
			if attribute.Value != dataType {
				return errors.Errorf("Unexpected %s :%s. Want %s", rwDataTypeKey, attribute.Value, dataType)
			}
			dataTypeValidated = true
		case rwBitLengthKey:
			if attribute.Value != fmt.Sprintf("%d", bitLength) {
				return errors.Errorf("Unexpected %s '%s'. Want '%d'", rwBitLengthKey, attribute.Value, bitLength)
			}
			bitLengthValidate = true
		}
	}
	if !dataTypeValidated {
		return errors.Errorf("required attribute %s missing", rwDataTypeKey)
	}
	if !bitLengthValidate {
		return errors.Errorf("required attribute %s missing", rwBitLengthKey)
	}
	return nil
}

File: plc4j/spi/buffers/xml/src/main/java/org/apache/plc4x/java/spi/buffers/xmlbased/ReadBufferXmlBased.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.spi.buffers.xmlbased;

import org.apache.plc4x.java.spi.buffers.api.AbstractBuffer;
import org.apache.plc4x.java.spi.buffers.api.ReadBuffer;
import org.apache.plc4x.java.spi.buffers.api.WithOption;
import org.apache.plc4x.java.spi.buffers.api.exceptions.BufferException;

import javax.xml.stream.XMLEventReader;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.events.*;
import java.io.InputStream;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Iterator;
import java.util.Optional;

public class ReadBufferXmlBased extends AbstractBuffer implements ReadBuffer, XmlBuffer {

    private final int sizeInBits;
    private int positionInBits;

    XMLEventReader reader;

    public ReadBufferXmlBased(InputStream is) {
        XMLInputFactory xmlInputFactory = XMLInputFactory.newInstance();
        xmlInputFactory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
        // TODO: Set this ....
        sizeInBits = 0;
        positionInBits = 0;
        try {
            reader = xmlInputFactory.createXMLEventReader(is);
        } catch (XMLStreamException e) {
            throw new IllegalArgumentException(e);
        }
    }

    @Override
    public void pushContext(WithOption... options) throws BufferException {
        String name = getName(options);
        StartElement startElement = travelToNextStartElement();
        String elementName = startElement.getName().getLocalPart();
        if (!elementName.equals(name)) {
            throw new BufferException(String.format("Unexpected start element '%s'. Expected '%s'", elementName, name));
        }
        super.pushContext(options);
    }

    @Override
    public void popContext(WithOption... options) throws BufferException {
        String name = getName(WithOption.AddOptions(getContext(), options));
        EndElement endElement = travelToNextEndElement();
        String elementName = endElement.getName().getLocalPart();
        if (!elementName.equals(name)) {
            throw new BufferException(String.format("Unexpected end element '%s'. Expected '%s'", elementName, name));
        }
        super.popContext(options);
    }

    @Override
    public boolean readBit(WithOption... options) throws BufferException {
        String name = getName(WithOption.AddOptions(getContext(), options));
        String bitString = decode(name, rwBitKey, 1);
        move(1);
        if (!"true".equalsIgnoreCase(bitString) && !"false".equalsIgnoreCase(bitString)) {
            throw new BufferException(String.format("Unexpected bit value '%s'. Expected 'true' or 'false'", bitString));
        }
        return Boolean.parseBoolean(bitString);
    }

    @Override
    public byte[] readBits(int numBits, WithOption... options) throws BufferException {
        String name = getName(options);
        String hexString = decode(name, rwByteKey, numBits);
        if (!hexString.startsWith("0x")) {
            throw new BufferException(String.format("Hex string should start with 0x. Actual value %s", hexString));
        }
        hexString = hexString.substring(2);
        int numberOfBytes = (numBits + 7) / 8;
        if (hexString.length() != numberOfBytes * 2) {
            throw new BufferException(String.format("Hex string should be %d bytes long. Actual value %s", numberOfBytes, hexString));
        }
        byte[] bytes = new byte[numberOfBytes];
        for (int i = 0; i < hexString.length(); i = i + 2) {
            // Without this hack, we can't parse values such as "0x80" into a byte.
            bytes[i / 2] = (byte) Short.parseShort(hexString.substring(i, i + 2), 16);
        }
        return bytes;
    }

    @Override
    public byte readUnsignedByte(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Byte.parseByte(decode(name, rwUintKey, numBits));
    }

    @Override
    public short readUnsignedShort(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Short.parseShort(decode(name, rwUintKey, numBits));
    }

    @Override
    public int readUnsignedInt(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Integer.parseInt(decode(name, rwUintKey, numBits));
    }

    @Override
    public long readUnsignedLong(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Long.parseLong(decode(name, rwUintKey, numBits));
    }

    @Override
    public BigInteger readUnsignedBigInteger(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        String bigIntString = decode(name, rwUintKey, numBits);
        return new BigInteger(bigIntString);
    }

    @Override
    public byte readSignedByte(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Byte.parseByte(decode(name, rwIntKey, numBits));
    }

    @Override
    public short readSignedShort(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Short.parseShort(decode(name, rwIntKey, numBits));
    }

    @Override
    public int readSignedInt(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Integer.parseInt(decode(name, rwIntKey, numBits));
    }

    @Override
    public long readSignedLong(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Long.parseLong(decode(name, rwIntKey, numBits));
    }

    @Override
    public BigInteger readSignedBigInteger(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return new BigInteger(decode(name, rwIntKey, numBits));
    }

    @Override
    public float readFloat(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Float.parseFloat(decode(name, rwFloatKey, numBits));
    }

    @Override
    public double readDouble(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return Double.parseDouble(decode(name, rwFloatKey, numBits));
    }

    @Override
    public BigDecimal readBigDecimal(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return new BigDecimal(decode(name, rwFloatKey, numBits));
    }

    @Override
    public String readString(int numBits, WithOption... options) throws BufferException {
        String name = getScalarName(options);
        move(numBits);
        return decode(name, rwStringKey, numBits);
    }

    @Override
    public ReadBuffer createSubBuffer(int numBits, WithOption... options) {
        // TODO: Find out how to do this.
        return null;
    }

    @Override
    public int getPositionInBits() {
        return positionInBits;
    }

    @Override
    public int getRemainingBits() {
        return sizeInBits - positionInBits;
    }

    @Override
    public void setPositionInBits(int positionInBits) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    private void move(int bits) {
        positionInBits += bits;
    }

    private StartElement travelToNextStartElement() throws BufferException {
        while (reader.hasNext()) {
            XMLEvent xmlEvent;
            try {
                xmlEvent = reader.nextEvent();
            } catch (XMLStreamException e) {
                throw new BufferException("Error reading next element", e);
            }
            if (xmlEvent.isStartElement()) {
                return xmlEvent.asStartElement();
            } else if (xmlEvent.isEndElement()) {
                throw new BufferException(String.format("Unexpected End element '%s'", xmlEvent.asEndElement().getName().getLocalPart()));
            }
        }
        throw new BufferException("EOF");
    }

    private EndElement travelToNextEndElement() throws BufferException {
        while (reader.hasNext()) {
            XMLEvent xmlEvent;
            try {
                xmlEvent = reader.nextEvent();
            } catch (XMLStreamException e) {
                throw new BufferException("Error reading next element", e);
            }
            if (xmlEvent.isStartElement()) {
                throw new BufferException(String.format("Unexpected Start element '%s'", xmlEvent.asStartElement().getName().getLocalPart()));
            } else if (xmlEvent.isEndElement()) {
                return xmlEvent.asEndElement();
            }
        }
        throw new BufferException("EOF");
    }

    private String decode(String name, String dataType, int numBits) throws BufferException {
        StartElement startElement = travelToNextStartElement();
        validateStartElement(startElement, name, dataType, numBits);
        Characters characters;
        try {
            characters = reader.nextEvent().asCharacters();
        } catch (XMLStreamException e) {
            throw new BufferException("Error reading characters.", e);
        }
        String data = characters.getData();
        try {
            XMLEvent xmlEvent = reader.nextEvent();
            if (!xmlEvent.isEndElement()) {
                throw new BufferException("Expected end element.");
            }
        } catch (XMLStreamException e) {
            throw new BufferException("Error reading end element.", e);
        }
        return data.trim();
    }

    private void validateStartElement(StartElement startElement, String name, String dataType, int numBits) throws BufferException {
        //name = sanitizeLogicalName(name);
        String elementName = startElement.getName().getLocalPart();
        // "*" matches any element name: used for elements whose serialized name is only known
        // dynamically (e.g. enum fields are written as <fieldName><EnumTypeName ...>).
        if (!"*".equals(name) && !elementName.equals(name) && !isReservedFieldName(name, elementName)) {
            throw new BufferException(String.format("unexpected element '%s'. Expected '%s'", elementName, name));
        }
        validateAttr(startElement.getAttributes(), name, dataType, numBits);
    }

    private boolean isReservedFieldName(String expectedName, String elementName) {
        // The generated parsers qualify reserved fields as "<Type>.reserved<N>" while the
        // generated serializers write them plainly as "reserved" - accept the serializer's
        // form so XML written by WriteBufferXmlBased can be read back.
        return "reserved".equals(elementName) && expectedName.matches("(.*\\.)?reserved\\d*");
    }

    private void validateAttr(Iterator<Attribute> attr, String name, String dataType, int numBits) throws BufferException {
        boolean dataTypeValidated = false;
        boolean numBitsValidate = false;
        while (attr.hasNext()) {
            Attribute attribute = attr.next();
            if (attribute.getName().getLocalPart().equals(rwDataTypeKey)) {
                if (!attribute.getValue().equals(dataType)) {
                    throw new BufferException(String.format("%s: Unexpected dataType '%s'. Want '%s'", name, attribute.getValue(), dataType));
                }
                dataTypeValidated = true;
            } else if (attribute.getName().getLocalPart().equals(rwBitLengthKey)) {
                if (!attribute.getValue().equals(Integer.toString(numBits))) {
                    throw new BufferException(String.format("%s: Unexpected numBits '%s'. Want '%d'", name, attribute.getValue(), numBits));
                }
                numBitsValidate = true;
            }
        }
        if (!dataTypeValidated) {
            throw new BufferException(String.format("%s: required attribute '%s' missing", name, rwDataTypeKey));
        }
        if (!numBitsValidate) {
            throw new BufferException(String.format("%s: required attribute '%s' missing", name, rwBitLengthKey));
        }
    }

    /**
     * Resolves the element name for a scalar read.
     *
     * <p>Mirrors {@code FieldWriterArray.writeSimpleTypeArrayField}, which serializes every simple
     * list item as {@code <value>} and discards the field's own name. Reads carrying the
     * render-as-list marker are exactly those items, so without this a list serialized as
     * {@code <classId isList="true"><value/>...} fails to read back with
     * "unexpected element 'value'. Expected 'classId'". Only scalar reads go through here:
     * byte arrays are written as a single element, and complex list items push their own type
     * name as context.
     */
    private String getScalarName(WithOption... options) throws BufferException {
        if (WithOption.extractRenderAsList(options).orElse(false)) {
            return "value";
        }
        return getName(options);
    }

    protected String getName(WithOption... options) throws BufferException {
        Optional<String> name = WithOption.extractName(options, getContext());
        if (name.isEmpty()) {
            throw new BufferException("Missing 'name' option.");
        }
        return name.get();
    }

}
